### ISSUE TITLE
ring buffer data structure

### ISSUE THREAD

**Issue by twood02** [gwAdvNet2015/adv-net-samples#2]
Write a ring buffer data structure in `datastructs/ring.c`.  The [header file](https://github.com/gwAdvNet2015/adv-net-samples/blob/master/datastructs/ring.h) provides an outline.

```
/* A ring buffer for multi-threaded programs. A single producer pushes
* objects into the ring, while a single consumer pops them off for
* processing. If the ring is full (i.e., the consumer is slower than the
* producer), then the push operation fails.
*/
```

Feel free to modify the declarations in the header file if you have a better idea.

**Comment by wenhuizhang** [comment]
@huyang1988  @pradeep-k @wenhuizhang

**Comment by huyang1988** [comment]
Hi

**Comment by twood02** [comment]
code merged in #26.
